### PR TITLE
Added gRPC support for winch. Added OIDC support for gRPC itself.

### PR DESCRIPTION
**Pull request by Bplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: lib/discovery/updater_test.go [hunk 1]
```diff
@@ -143,27 +143,29 @@ func TestUpdater_OnEvent_AdditionAndModifyAndDelete_HTTP(t *testing.T) {
 				{
 					Autogenerated: true,
 					HostMatcher:   "s2.external2.host.com",
-					PortMatcher:   222,
+					PortMatcher:   80,
```
Comment (kdima) [inline lib/discovery/updater_test.go]: i gave up on this whole file. We should think about refactoring this somehow. It is super hard to follow or figure out what is going ion

File: server/main.go [hunk 2]
```diff
@@ -162,11 +146,36 @@ func main() {
...
+		grpc_logrus.UnaryServerInterceptor(logEntry),
+		grpc_prometheus.UnaryServerInterceptor,
+	}
+	grpcStreamInterceptors := []grpc.StreamServerInterceptor{
+		grpc_ctxtags.StreamServerInterceptor(),
+		grpc_logrus.StreamServerInterceptor(logEntry),
+		grpc_prometheus.StreamServerInterceptor,
+	}
+
 	if authorizer != nil {
```
Comment (kdima) [inline server/main.go]: can authorizer ever be nil? Or shall we log warn or err when that happens or even panic?

Comment (Bplotka) [reply server/main.go]: yes, it can be nil when user will specify no authorization at all (e.g just cert authorization)

File: http/director/router/router_test.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+		{
+			scheme: "http",
+			host:   "nopath.example.com",
+			port:   "80",
+			path:   "/test/path",
+
+			expectedBackend: "a",
+		},
+		{
+			scheme: "http",
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply http/director/router/router_test.go]: Because port matcher on "a" is 80, but this guy has 83 so it ends up in generic rule to "b"

File: http/integration_test.go [hunk 4]
```diff
@@ -642,17 +649,30 @@ func (s *HttpProxyingIntegrationSuite) TestCallOverClient() {
...
 
 func (s *HttpProxyingIntegrationSuite) TestCallOverClient_WithPort_MatchWithGenericRoute() {
-	req := testRequest("http://nonsecure.ext.example.com:80/some/strict/path", "bearer abc8", testProxyAuthValue)
+	req := testRequest("http://nonsecure.ext.example.com:845350/some/strict/path", "bearer abc8", testProxyAuthValue)
 	resp, err := s.kedgeClient.Do(req)
 	s.assertSuccessfulPingback(req, resp, "bearer abc8", err)
 }
 
 func (s *HttpProxyingIntegrationSuite) TestCallOverClient_WithPort_SpecialRoute() {
+	fmt.Println("lol")
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply http/integration_test.go]: OMG

File: lib/discovery/updater_test.go (hunk 1 above)
Comment (Bplotka) [reply lib/discovery/updater_test.go]: hm, I separated these into two different files for grpc and http...but not sure how we can make it simpler or more clean

File: http/integration_test.go [hunk 5]
```diff
@@ -642,17 +649,30 @@ func (s *HttpProxyingIntegrationSuite) TestCallOverClient() {
 }
 
 func (s *HttpProxyingIntegrationSuite) TestCallOverClient_WithPort_MatchWithGenericRoute() {
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply http/integration_test.go]: Added
